File: screenshot-testing-core/src/main/java/com/facebook/testing/screenshot/internal/AlbumImpl.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.testing.screenshot.internal;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

import io.github.usefulness.testing.screenshot.ReportArtifactsManager;
import io.github.usefulness.testing.screenshot.ScreenshotDirectories;

/**
 * A "local" implementation of Album.
 */
public class AlbumImpl implements Album {
    private static final int COMPRESSION_QUALITY = 90;

    private final Set<String> mAllNames = new HashSet<>();
    private final MetadataRecorder mMetadataRecorder;
    private final ReportArtifactsManager mReportArtifactsManager;

    /* VisibleForTesting */
    AlbumImpl(ScreenshotDirectories screenshotDirectories) {
        mMetadataRecorder = new MetadataRecorder(screenshotDirectories);
        mReportArtifactsManager = new ReportArtifactsManager(screenshotDirectories);
    }

    /**
     * Creates a "local" album that stores all the images on device.
     */
    public static AlbumImpl create() {
        return new AlbumImpl(new ScreenshotDirectories());
    }

    @Override
    public void flush() {
        mMetadataRecorder.flush();
    }

    @Override
    public String writeBitmap(String name, int tilei, int tilej, Bitmap bitmap) {
        String tileName = generateTileName(name, tilei, tilej);
        String filename = getScreenshotFilenameInternal(tileName);
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.PNG, COMPRESSION_QUALITY, os);
        mReportArtifactsManager.recordFile(filename, os.toByteArray());
        return tileName;
    }

    /**
     * Same as the public getScreenshotFile() except it returns the File even if the screenshot
     * doesn't exist.
     */
    private static String getScreenshotFilenameInternal(String name) {
        return name + ".png";
    }

    private static String getViewHierarchyFilename(String name) {
        return name + "_dump.json";
    }

    private static String getAxIssuesFilename(String name) {
        return name + "_issues.json";
    }

    @Override
    public void writeAxIssuesFile(String name, String data) {
        writeMetadataFile(getAxIssuesFilename(name), data);
    }

    @Override
    public void writeViewHierarchyFile(String name, String data) {
        writeMetadataFile(getViewHierarchyFilename(name), data);
    }

    public void writeMetadataFile(String name, String data) {
        byte[] out = data.getBytes();
        mReportArtifactsManager.recordFile(name, out);
    }

    /**
     * Add the given record to the album. This is called by RecordBuilderImpl#record() and so is an
     * internal detail.
     */
    @SuppressLint("SetWorldReadable")
    @Override
    public void addRecord(RecordBuilderImpl recordBuilder) throws IOException {
        recordBuilder.checkState();
        if (mAllNames.contains(recordBuilder.getName())) {
            if (recordBuilder.hasExplicitName()) {
                throw new AssertionError(
                    "Can't create multiple screenshots with the same name: " + recordBuilder.getName());
            }

            throw new AssertionError(
                "Can't create multiple screenshots from the same test, or "
                    + "use .setName() to name each screenshot differently");
        }

        Tiling tiling = recordBuilder.getTiling();

        MetadataRecorder.ScreenshotMetadataRecorder screenshotNode =
            mMetadataRecorder
                .addNewScreenshot()
                .withDescription(recordBuilder.getDescription())
                .withName(recordBuilder.getName())
                .withTestClass(recordBuilder.getTestClass())
                .withTestName(recordBuilder.getTestName())
                .withTileWidth(tiling.getWidth())
                .withTileHeight(tiling.getHeight())
                .withViewHierarchy(getViewHierarchyFilename(recordBuilder.getName()))
                .withAxIssues(getAxIssuesFilename(recordBuilder.getName()))
                .withExtras(recordBuilder.getExtras());

        if (recordBuilder.getError() != null) {
            screenshotNode.withError(recordBuilder.getError());
        }

        if (recordBuilder.getGroup() != null) {
            screenshotNode.withGroup(recordBuilder.getGroup());
        }

        mAllNames.add(recordBuilder.getName());

        screenshotNode.save();
    }

    /**
     * For a given screenshot, and a tile position, generates a name where we store the screenshot in
     * the album.
     *
     * <p>For backward compatibility with existing screenshot scripts, for the tile (0, 0) we use the
     * name directly.
     */
    private String generateTileName(String name, int i, int j) {
        if (i == 0 && j == 0) {
            return name;
        }

        return String.format("%s_%s_%s", name, i, j);
    }
}

File: screenshot-testing-core/src/main/kotlin/io/github/usefulness/testing/screenshot/internal/AlbumImpl.kt
package io.github.usefulness.testing.screenshot.internal

import android.graphics.Bitmap

internal class AlbumImpl : Album {

    override fun writeBitmap(name: String, tilei: Int, tilej: Int, bitmap: Bitmap): String {
        val tileName = generateTileName(name, tilei, tilej)
        val filename = getScreenshotFilenameInternal(tileName)
        ScreenshotDirectories.openOutputFile(filename).use {
            bitmap.compress(Bitmap.CompressFormat.PNG, COMPRESSION_QUALITY, it)
        }
        return tileName
    }

    override fun writeAxIssuesFile(name: String, data: String) {
        writeMetadataFile(getAxIssuesFilename(name), data)
    }

    override fun writeViewHierarchyFile(name: String, data: String) {
        writeMetadataFile(getViewHierarchyFilename(name), data)
    }

    private fun writeMetadataFile(name: String, data: String) {
        ScreenshotDirectories.openOutputFile(name)
            .use { it.write(data.toByteArray()) }
    }

    override fun addRecord(recordBuilder: RecordBuilderImpl) {
        recordBuilder.checkState()
        if (MetadataRecorder.snapshot().any { it.name == recordBuilder.name }) {
            if (recordBuilder.hasExplicitName()) {
                error("Can't create multiple screenshots with the same name: ${recordBuilder.name}")
            }

            error("Can't create multiple screenshots from the same test, or use .setName() to name each screenshot differently")
        }

        val tiling = recordBuilder.tiling

        MetadataRecorder.addNew(
            screenshot = MetadataRecorder.ScreenshotMetadata(
                description = recordBuilder.description,
                name = recordBuilder.name,
                testClass = recordBuilder.testClass,
                testName = recordBuilder.testName,
                tileWidth = tiling.width,
                tileHeight = tiling.height,
                viewHierarchy = getViewHierarchyFilename(recordBuilder.name),
                axIssues = getAxIssuesFilename(recordBuilder.name),
                extras = recordBuilder.extras,
                error = recordBuilder.error,
                group = recordBuilder.group,
            ),
        )
    }

    companion object {

        private const val COMPRESSION_QUALITY = 90

        /**
         * For a given screenshot, and a tile position, generates a name where we store the screenshot in
         * the album.
         *
         *
         * For backward compatibility with existing screenshot scripts, for the tile (0, 0) we use the
         * name directly.
         */
        private fun generateTileName(name: String, i: Int, j: Int): String {
            if (i == 0 && j == 0) {
                return name
            }

            return "${name}_${i}_$j"
        }

        private fun getScreenshotFilenameInternal(name: String?): String = "$name.png"

        private fun getViewHierarchyFilename(name: String?): String = name + "_dump.json"

        private fun getAxIssuesFilename(name: String?): String = name + "_issues.json"
    }
}
